node/core/backing/Cargo.toml

```diff
@@ -22,7 +23,9 @@ log = "0.4.8"
 
 [dev-dependencies]
 sp-core = { git = "https://github.com/paritytech/substrate", branch = "master" }
+sp-application-crypto = { git = "https://github.com/paritytech/substrate", branch = "master" }
 sp-keyring = { git = "https://github.com/paritytech/substrate", branch = "master" }
 futures = { version = "0.3.5", features = ["thread-pool"] }
+tokio = "0.2.22"
```

Comment:
why do we need tokio here?

Reply:
Added it to use Tokio's `block_on`. With `futures::executor`, calling `block_on`(in the test for example) and downstream also calling `block_on` (in the keystore) would panic because you cannot have multiple `LocalPool`s running. @tomaka once suggested using a different runtime to workaround this issue and that's what i did here.

Reply:
Sorry, I don't follow. Are you referring to nested `block_on` calls leads to a panic? I don't see it happening here. `TestState::default` is never called within `block_on` block as far as I can see.

The proper solution to this problem is to avoid having nested `block_on` calls by making some functions async.

Reply:
[Here's](https://github.com/paritytech/polkadot/pull/1740/files#diff-b2e256a6d7ea7397ed916a1ed6d74934L1002) where `block_on` happens. Downstream, the keystore is going to be called synchronously which is going to use `block_on` again that causes the panic. This is why using a different runtime to avoid this in tests is required. 

> The proper solution to this problem is to avoid having nested block_on calls by making some functions async.

That's correct. We are going to work on follow up PRs where parts of substrate / polkadot are going to be converted to async/await so that we can avoid using `SyncCryptoStore` altogether.

Reply:
> Downstream, the keystore is going to be called synchronously

but what is calling it exactly?

I applied the following patch to #1746 and `cargo test -p polkadot-node-core-backing` passes:
```patch
diff --git a/Cargo.lock b/Cargo.lock
index 65d8b86b..ee8974d4 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -4982,7 +4982,6 @@ dependencies = [
  "sp-core",
  "sp-keyring",
  "sp-keystore",
- "tokio 0.2.22",
 ]
 
 [[package]]
diff --git a/node/core/backing/Cargo.toml b/node/core/backing/Cargo.toml
index f2efdf2d..0e7a0a5f 100644
--- a/node/core/backing/Cargo.toml
+++ b/node/core/backing/Cargo.toml
@@ -26,6 +26,5 @@ log = "0.4.8"
 sp-application-crypto = { git = "https://github.com/rakanalh/substrate", branch = "async-keystore-auth-discovery" }
 sp-keyring = { git = "https://github.com/rakanalh/substrate", branch = "async-keystore-auth-discovery" }
 futures = { version = "0.3.5", features = ["thread-pool"] }
-tokio = "0.2.22"
 assert_matches = "1.3.0"
 polkadot-node-subsystem-test-helpers = { path = "../../subsystem-test-helpers" }
diff --git a/node/core/backing/src/lib.rs b/node/core/backing/src/lib.rs
index c15041ba..13d089a2 100644
--- a/node/core/backing/src/lib.rs
+++ b/node/core/backing/src/lib.rs
@@ -887,7 +887,6 @@ mod tests {
 	use sp_application_crypto::AppKey;
 	use sp_keystore::CryptoStore;
 	use std::collections::HashMap;
-	use tokio::runtime::Runtime;
 
 	fn validator_pubkeys(val_ids: &[Sr25519Keyring]) -> Vec<ValidatorId> {
 		val_ids.iter().map(|v| v.public().into()).collect()
@@ -1008,8 +1007,7 @@ mod tests {
 
 		futures::pin_mut!(test_fut);
 		futures::pin_mut!(subsystem);
-		let mut rt = Runtime::new().unwrap();
-		rt.block_on(future::select(test_fut, subsystem));
+		futures::executor::block_on(future::select(test_fut, subsystem));
 	}
 
 	fn make_erasure_root(test: &TestState, pov: PoV) -> Hash {
```